### PR TITLE
retrieve encryption key

### PR DESCRIPTION
**Pull request by jgsqware**
Allow to retreive encryption key for k8s secret

Related: https://github.com/giantswarm/giantswarm/issues/1907

Signed-off-by: Julien Garcia Gonzalez <julien@giantswarm.io>

### CODE REVIEW

**Comment by xh3b4sd** [comment]
AFAIK we agreed on putting this into a new repository. Why did you add the code here now?

**Comment by jgsqware** [comment]
quick and dirty fix to move on k8s benchmark security

If it really need to move this now on `randomkeytpr` I would do it

File: service.go [hunk 1]
```diff
@@ -143,3 +149,57 @@ func (s *Service) SearchCertsForComponent(clusterID, componentName string) (Asse
...
+				return "", microerror.Maskf(secretsRetrievalFailedError, "secrets channel was already closed")
+			}
+
+			switch event.Type {
+			case watch.Added:
+				secret := event.Object.(*v1.Secret)
+				// component := ClusterComponent(secret.Labels[ComponentLabel])
+				// if !ValidComponent(component, ClusterComponents) {
+				// 	return nil, microerror.Maskf(secretsRetrievalFailedError, "unknown clusterComponent %s", component)
+				// }
```
Comment (xh3b4sd) [inline service.go:187]: Why is this commented?

File: service.go [hunk 2]
```diff
@@ -143,3 +149,57 @@ func (s *Service) SearchCertsForComponent(clusterID, componentName string) (Asse
...
+
+				asset, ok := secret.Data["encryption"]
+				if !ok {
+					return "", microerror.Maskf(secretsRetrievalFailedError, "malformed secret was missing %v asset", "encryption")
+				}
+
+				return string(asset[:]), nil
```
Comment (xh3b4sd) [inline service.go:194]: AFAIK this creates a copy of the byte slice. Why are we doing this?

File: service.go [hunk 3]
```diff
@@ -83,6 +83,12 @@ func (s *Service) SearchCerts(clusterID string) (AssetsBundle, error) {
 	return assetsBundle, nil
 }
 
+// SearchEncryptionKey watches for encruption key secrets of a cluster  and returns it as
+// assets bundle.
+func (s *Service) SearchEncryptionKey(clusterID string) (string, error) {
+	return s.SearchEncryptionKeyForComponent(clusterID, "k8s-encryption")
```
Comment (xh3b4sd) [inline service.go:89]: Is there now no conflict because we use a different name?

Comment (jgsqware) [reply service.go:89]: yes exactly

**Review by rossf7 — COMMENTED** [review]
Late to the party here but I feel like I'm missing the full picture.

Why was the encryption key added to `clustertpr` as well as watching for the secret? Can we not just get the encryption key from the secret?

**Comment by xh3b4sd** [comment]
> Why was the encryption key added to clustertpr

This is actually a good question. I think we can revert this change since there should just be a file added to the cloud config via extensions like for the certificates.

**Comment by jgsqware** [comment]
I've added it the clusterTPR for two reason, First it was a genric option for the api as the port, etc... so I found useful to have it there. Second one probably for easiness, because the templatisation use clusterTPR as object. 
But I probably missed the scope of clusterTPR

**Comment by xh3b4sd** [comment]
The odd thing is that the new field is then always empty in the TPO in the k8s API and that the value is only set for templating. I saw in k8scloudconfig that you provide a file path anyway so adding a file via extension would be the right thing and there would be no need for this weird placeholder in the clustertpr. We already have the Vault token in the clustertpr which is wrong and we should add it via files or something as well.

**Comment by jgsqware** [comment]
Ok noted, I will revert that

**Comment by jgsqware** [comment]
However, the current implementation create the file via this https://github.com/giantswarm/aws-operator/blob/encrypt-k8s-at-rest/service/cloudconfig/templates.go#L148. 
Is it that the `extension` you mean? 
If yes, the template values comes from `awstpr.Spec`, it's why I finally used the `clustertrp`.
I'm agree with you POV about clustertpr but from where do the data come from for template?

**Comment by rossf7** [comment]
The templating using the cluster object isn't going to work as the key will always be blank. It is set by kubernetesd when the cluster object is created.

I think you'll need to render the template in the operator using the key you get from the secret. You can then embed the unit including the key value in the cloudconfig.

**Comment by xh3b4sd** [comment]
This is what you are looking for: https://github.com/giantswarm/aws-operator/blob/master/service/cloudconfig/v_0_1_0_master_template.go.

**Comment by jgsqware** [comment]
Close in favor of https://github.com/giantswarm/randomkeytpr/pull/1

**Pull request closed without merging**